flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveTableLookupFunction.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.flink.table.functions.TableFunction;
+import org.apache.flink.table.runtime.typeutils.RowDataTypeInfo;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Lookup table function for Hive tables.
+ */
+public class HiveTableLookupFunction extends TableFunction<RowData> {
+
+	private static final long serialVersionUID = 1L;
+
+	private final HiveTableInputFormat inputFormat;
+	// indices of lookup columns in the record returned by input format
+	private final int[] lookupCols;
+	private transient Map<RowData, List<RowData>> cache;
+	// timestamp when cache expires
+	private transient long cacheExpire;
+	private final Duration cacheTTL = Duration.ofHours(1);
+
+	public HiveTableLookupFunction(HiveTableInputFormat inputFormat, String[] lookupKeys) {
+		lookupCols = new int[lookupKeys.length];
+		String[] allFields = inputFormat.getFieldNames();
+		Map<String, Integer> nameToIndex = IntStream.range(0, allFields.length).boxed().collect(
+				Collectors.toMap(i -> allFields[i], i -> i));
+		List<Integer> selectedIndices = Arrays.stream(inputFormat.getSelectedFields()).boxed().collect(Collectors.toList());
+		for (int i = 0; i < lookupKeys.length; i++) {
+			Integer index = nameToIndex.get(lookupKeys[i]);
+			Preconditions.checkArgument(index != null, "Lookup keys %s not found in table schema", Arrays.toString(lookupKeys));
+			index = selectedIndices.indexOf(index);
+			Preconditions.checkArgument(index >= 0, "Lookup keys %s not selected", Arrays.toString(lookupKeys));
+			lookupCols[i] = index;
+		}
+		this.inputFormat = inputFormat;
+	}
+
+	@Override
+	public TypeInformation<RowData> getResultType() {
+		String[] allNames = inputFormat.getFieldNames();
+		DataType[] allTypes = inputFormat.getFieldTypes();
+		int[] selected = inputFormat.getSelectedFields();
+		return new RowDataTypeInfo(
+				Arrays.stream(selected).mapToObj(i -> allTypes[i].getLogicalType()).toArray(LogicalType[]::new),
+				Arrays.stream(selected).mapToObj(i -> allNames[i]).toArray(String[]::new));
+	}
+
+	@Override
+	public void open(FunctionContext context) throws Exception {
+		super.open(context);
+		cache = new HashMap<>();
+		cacheExpire = -1;
+	}
+
+	public void eval(Object... values) {
+		Preconditions.checkArgument(values.length == lookupCols.length, "Number of values and lookup keys mismatch");
+		reloadCache();
+		RowData probeKey = GenericRowData.of(values);
+		List<RowData> matchedRows = cache.get(probeKey);
+		if (matchedRows != null) {
+			for (RowData matchedRow : matchedRows) {
+				collect(matchedRow);
+			}
+		}
+	}
+
+	private void reloadCache() {
```

Comment:
does it mean it caches all rows of a hive file?

Reply:
Not all rows of a hive file, but all rows of a hive table. If we only cache part of the data, we have to do full table scan whenever we have a cache miss. So I guess we might as well cache all the data.

Reply:
I'm really concerned about caching the whole table, especially every parallelism does so.

We'd better add explicit documentation to warn users potential consequences and best practices.

Reply:
And your concern is valid. We will add explicit doc for that.
BTW, in our internal implementation, we can do a partitioned lookup join, so that each task only has to cache data that belongs to a specific shuffle partition which reduces memory consumption. But I don' think that optimization can make it for 1.11.

Reply:
@lirui-apache thanks for the explanation!